yourbench/pipeline/summarization.py

```diff
@@ -0,0 +1,497 @@
+        return 0.0
+
+
+def _run_inference_with_timeout(
+    config: Dict[str, Any], inference_calls: List[InferenceCall], stage_name: str, timeout_seconds: float
+) -> Optional[Dict[str, List[str]]]:
+    """
+    Run inference with a forced timeout, preventing infinite hang.
+
+    Args:
+        config (Dict[str, Any]): Pipeline configuration dictionary.
+        inference_calls (List[InferenceCall]): A list of calls to be passed to run_inference.
+        stage_name (str): The pipeline stage name (e.g. "summarization").
+        timeout_seconds (float): Timeout in seconds before we consider it a failure.
+
+    Returns:
+        Optional[Dict[str, List[str]]]: Dictionary of responses per model. If
+        timed out or errored, returns None.
+    """
+    logger.info("Attempting inference with a maximum timeout of {} seconds...", timeout_seconds)
+    with ThreadPoolExecutor(max_workers=1) as executor:
+        future = executor.submit(run_inference, config, stage_name, inference_calls)
+        try:
+            result = future.result(timeout=timeout_seconds)
+            if result is None or not isinstance(result, dict):
+                logger.error("Inference returned None or invalid result type: {}", type(result))
+                return None
+
+            # Check if the result dictionary is empty
+            if not result:
+                logger.error("Inference returned an empty dictionary")
+                return None
+
+            # Check if any model returned empty response list
+            for model_name, responses in result.items():
+                if not responses:
+                    logger.warning("Model '{}' returned empty response list", model_name)
+                else:
+                    logger.info("Received {} responses from model '{}'", len(responses), model_name)
+
+            return result
+
+        except FuturesTimeoutError:
+            logger.error("Inference timed out after {} seconds.", timeout_seconds)
+        except Exception as exc:
+            logger.error("Error during inference: {}", str(exc))
+
+    return None
+
+
+def duplicate_rows(dataset: Dict[str, Any], num_duplicates: int = 1) -> Dict[str, List[Any]]:
+    """
+    Create a dictionary that repeats each value in the dataset multiple times.
+
+    Args:
+        dataset (Dict[str, Any]): A dictionary representing dataset columns.
+        num_duplicates (int): How many times to duplicate each row.
+
+    Returns:
+        Dict[str, List[Any]]: A new dictionary where each key's list is repeated
+        num_duplicates times.
+    """
+    # Example usage: repeat dataset rows for augmentation or testing.
+    # Not actively used in summarization, but provided to preserve functionality.
+    repeated_data = {}
+    for key, value in dataset.items():
+        repeated_data[key] = [val for val in value for _ in range(num_duplicates)]
+    return repeated_data
+
+
+def run(config: Dict[str, Any]) -> None:
```

Comment:
same as the other files, I would split this big function into individual functions, that way, if you need to change a single step, it's easier to manage without risking breaking the full pipe